src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java

```diff
@@ -0,0 +1,496 @@
+   *
+   * @param productIds the product ids to find a key mapping for.
+   * @param categoryIds the category ids to find a key mapping for.
+   * @param productTypeIds the productType ids to find a key mapping for.
+   * @param customObjectIds the custom object ids to find a key mapping for.
+   * @return a map of id to key representing products, categories, productTypes and customObjects in
+   *     the CTP project defined by the injected {@code ctpClient}.
+   */
+  @Nonnull
+  CompletableFuture<Map<String, String>> getIdToKeys(
+      @Nonnull final Set<String> productIds,
+      @Nonnull final Set<String> categoryIds,
+      @Nonnull final Set<String> productTypeIds,
+      @Nonnull final Set<String> customObjectIds) {
+
+    final Set<String> nonCachedProductIds = getNonCachedReferenceIds(productIds);
+    final Set<String> nonCachedCategoryIds = getNonCachedReferenceIds(categoryIds);
+    final Set<String> nonCachedProductTypeIds = getNonCachedReferenceIds(productTypeIds);
+    final Set<String> nonCachedCustomObjectIds = getNonCachedReferenceIds(customObjectIds);
+
+    if (nonCachedProductIds.isEmpty()
+        && nonCachedCategoryIds.isEmpty()
+        && nonCachedProductTypeIds.isEmpty()) {
+      return fetchCustomObjectKeys(customObjectIds);
+    }
+
+    List<List<String>> productIdsChunk = ChunkUtils.chunk(nonCachedProductIds, CHUNK_SIZE);
+    List<List<String>> categoryIdsChunk = ChunkUtils.chunk(nonCachedCategoryIds, CHUNK_SIZE);
+    List<List<String>> productTypeIdsChunk = ChunkUtils.chunk(nonCachedProductTypeIds, CHUNK_SIZE);
+
+    List<ResourceIdsGraphQlRequest> collectedRequests = new ArrayList<>();
+
+    collectedRequests.addAll(
+        createResourceIdsGraphQlRequests(productIdsChunk, GraphQlQueryResources.PRODUCTS));
+    collectedRequests.addAll(
+        createResourceIdsGraphQlRequests(categoryIdsChunk, GraphQlQueryResources.CATEGORIES));
+    collectedRequests.addAll(
+        createResourceIdsGraphQlRequests(productTypeIdsChunk, GraphQlQueryResources.PRODUCT_TYPES));
+
+    return ChunkUtils.executeChunks(getCtpClient(), collectedRequests)
+        .thenApply(ChunkUtils::flattenGraphQLBaseResults)
+        .thenApply(
+            results -> {
+              cacheResourceReferenceKeys(results);
+              return referenceIdToKeyCache;
+            })
+        .thenCompose(ignored -> fetchCustomObjectKeys(nonCachedCustomObjectIds));
+  }
+
+  @Nonnull
+  private CompletableFuture<Map<String, String>> fetchCustomObjectKeys(
+      @Nonnull final Set<String> nonCachedCustomObjectIds) {
+
+    if (nonCachedCustomObjectIds.isEmpty()) {
+      return CompletableFuture.completedFuture(referenceIdToKeyCache);
+    }
+
+    final List<List<String>> chunkedIds = ChunkUtils.chunk(nonCachedCustomObjectIds, CHUNK_SIZE);
+
+    final List<CustomObjectQuery<JsonNode>> chunkedRequests =
+        chunkedIds.stream()
+            .map(ids -> CustomObjectQuery.ofJsonNode().plusPredicates(p -> p.id().isIn(ids)))
```

Comment:
can you put withTotal=false and a limit to the query, otherwise it's limited to 20 custom objects, (we figured that with benchmark test in the previous release if you remember)

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/697/commits/c5e5c9b9c226f383f5d8ba70db7357c7bfdd0d8d